File: sdk-java/src/main/java/com/battlesim/naval/tactical/Helm.java
package com.battlesim.naval.tactical;

import com.battlesim.naval.Geometry;
import com.battlesim.naval.protocol.SelfState;
import com.battlesim.naval.protocol.ShipSpecs;
import com.battlesim.naval.protocol.Vec2;

/**
 * Steering helper.
 *
 * <p>Translates a desired bearing or waypoint into a {@code (throttle, rudder)}
 * pair that respects the speed-coupled turn rate of the ship. Sharp turns get
 * reduced throttle so the rudder bites rather than the hull plows straight.
 */
public class Helm {

    public static final class Config {
        public float mapWidth = 800.0f;
        public float mapHeight = 800.0f;
        public float wallMargin = 30.0f;
        public float turnAggressionDeg = 30.0f;
        public float alignThresholdDeg = 10.0f;
        public float minTurnThrottle = 0.55f;
    }

    /** Result of a steering decision. */
    public record Steering(float throttle, float rudder) {}

    private final ShipSpecs specs;
    private final Config cfg;

    public Helm(ShipSpecs specs) {
        this(specs, new Config());
    }

    public Helm(ShipSpecs specs, Config config) {
        this.specs = specs;
        this.cfg = config;
    }

    public Steering steerToBearing(SelfState me, float targetBearingDeg) {
        return steerToBearing(me, targetBearingDeg, true, 1.0f);
    }

    public Steering steerToBearing(
            SelfState me, float targetBearingDeg, boolean respectWalls, float desiredThrottle) {
        float bearing = respectWalls ? wallOverride(me, targetBearingDeg) : targetBearingDeg;
        float delta = Geometry.signedBearingDelta(bearing, me.headingDeg());
        float rudder = Geometry.clamp(delta / cfg.turnAggressionDeg, -1.0f, 1.0f);

        float absDelta = Math.abs(delta);
        float throttle;
        if (absDelta <= cfg.alignThresholdDeg) {
            throttle = desiredThrottle;
        } else {
            float denom = Math.max(180.0f - cfg.alignThresholdDeg, 1e-6f);
            float scale = Geometry.clamp((180.0f - absDelta) / denom, 0.0f, 1.0f);
            throttle = cfg.minTurnThrottle + (desiredThrottle - cfg.minTurnThrottle) * scale;
        }
        return new Steering(throttle, rudder);
    }

    public Steering steerToPoint(SelfState me, Vec2 target) {
        return steerToPoint(me, target, true, 1.0f);
    }

    public Steering steerToPoint(SelfState me, Vec2 target, boolean respectWalls, float desiredThrottle) {
        return steerToBearing(me, Geometry.bearingTo(me.pos(), target), respectWalls, desiredThrottle);
    }

    /**
     * If we're inside the wall margin and the target points further into the
     * wall, redirect toward an inward bearing instead.
     */
    float wallOverride(SelfState me, float targetBearing) {
        float x = me.pos().x();
        float y = me.pos().y();
        float pushX = 0.0f;
        float pushY = 0.0f;
        if (x < cfg.wallMargin) pushX = 1.0f;
        else if (x > cfg.mapWidth - cfg.wallMargin) pushX = -1.0f;
        if (y < cfg.wallMargin) pushY = 1.0f;
        else if (y > cfg.mapHeight - cfg.wallMargin) pushY = -1.0f;
        if (pushX == 0.0f && pushY == 0.0f) return targetBearing;

        float pushBearing = Geometry.bearingTo(new Vec2(0.0f, 0.0f), new Vec2(pushX, pushY));
        float delta = Math.abs(Geometry.signedBearingDelta(targetBearing, pushBearing));
        if (delta <= 90.0f) return targetBearing;
        return pushBearing;
    }
}

File: sdk-python/naval_sdk/tactical/helm.py
"""Steering helper.

Translates a desired bearing or waypoint into a ``(throttle, rudder)`` pair
that respects the speed-coupled turn rate of the ship (sharp turns need
less throttle, otherwise the rudder is wasted on a fast-but-straight hull).

See ``docs/design-decisions/sdk-tactical-toolkit.md`` §4.4.
"""

from __future__ import annotations

from typing import Tuple

from ..helpers import bearing_to, clamp, signed_bearing_delta
from ..protocol import SelfState, ShipSpecs


class Helm:
    def __init__(
        self,
        specs: ShipSpecs,
        *,
        map_width: float = 800.0,
        map_height: float = 800.0,
        wall_margin: float = 30.0,
        turn_aggression_deg: float = 30.0,
        align_threshold_deg: float = 10.0,
        min_turn_throttle: float = 0.55,
    ) -> None:
        self._specs = specs
        self._w = float(map_width)
        self._h = float(map_height)
        self._margin = float(wall_margin)
        self._turn_agg = float(turn_aggression_deg)
        self._align = float(align_threshold_deg)
        self._min_throttle = float(min_turn_throttle)

    def steer_to_bearing(
        self,
        me: SelfState,
        target_bearing_deg: float,
        *,
        respect_walls: bool = True,
        desired_throttle: float = 1.0,
    ) -> Tuple[float, float]:
        """Return ``(throttle, rudder)`` to align with ``target_bearing_deg``.

        Rudder is proportional to the signed bearing delta, scaled by
        ``turn_aggression_deg``. Throttle tapers toward ``min_turn_throttle``
        as the required turn grows, so the ship doesn't try to plough straight
        through a heading change.
        """
        bearing = (
            self._wall_override(me, target_bearing_deg) if respect_walls else target_bearing_deg
        )
        delta = signed_bearing_delta(bearing, me.heading_deg)
        rudder = clamp(delta / self._turn_agg, -1.0, 1.0)

        abs_delta = abs(delta)
        if abs_delta <= self._align:
            throttle = desired_throttle
        else:
            scale = (180.0 - abs_delta) / max(180.0 - self._align, 1e-6)
            scale = clamp(scale, 0.0, 1.0)
            throttle = self._min_throttle + (desired_throttle - self._min_throttle) * scale
        return throttle, rudder

    def steer_to_point(
        self,
        me: SelfState,
        target: Tuple[float, float],
        *,
        respect_walls: bool = True,
        desired_throttle: float = 1.0,
    ) -> Tuple[float, float]:
        """Convenience: steer toward a world-space point."""
        return self.steer_to_bearing(
            me,
            bearing_to(me.pos, target),
            respect_walls=respect_walls,
            desired_throttle=desired_throttle,
        )

    # -- Internals ---------------------------------------------------------

    def _wall_override(self, me: SelfState, target_bearing: float) -> float:
        """If we're inside the wall margin and the target points further into
        the wall, redirect toward an inward bearing instead.
        """
        x, y = me.pos
        push_x = 0.0
        push_y = 0.0
        if x < self._margin:
            push_x = 1.0
        elif x > self._w - self._margin:
            push_x = -1.0
        if y < self._margin:
            push_y = 1.0
        elif y > self._h - self._margin:
            push_y = -1.0
        if push_x == 0.0 and push_y == 0.0:
            return target_bearing

        # Compass bearing of the push vector.
        push_bearing = bearing_to((0.0, 0.0), (push_x, push_y))
        # If the target bearing already agrees with the push direction (within
        # 90°), leave it alone — we're heading away from the wall on our own.
        delta = abs(signed_bearing_delta(target_bearing, push_bearing))
        if delta <= 90.0:
            return target_bearing
        return push_bearing
